devices.go

```diff
@@ -172,10 +183,48 @@ func (d *deviceCache) SetDeviceByIdOpState(id string, os models.OperatingState)
 
 // Update updates the device in the cache and ensures that the
 // copy in Core Metadata is also updated.
-func (d *deviceCache) Update(id string) error {
+func (d *deviceCache) Update(dev *models.Device) error {
+	addressable, err := updateAddressable(&dev.Addressable)
+	if err != nil {
+		return err
+	}
+	dev.Addressable = *addressable
+
+	err = svc.dc.Update(*dev)
+	if err != nil {
+		return err
+	}
+
+	// consider device name can be modified, so remove the old one and put new one
+	if _, ok := d.names[dev.Id.Hex()]; ok {
+		delete(d.devices, d.names[dev.Id.Hex()])
+	}
+	d.devices[dev.Name] = dev
+	d.names[dev.Id.Hex()] = dev.Name
+
 	return nil
 }
 
+// updateAddressable updates the addressable.
+// Because addressable client doesn't have update function. If addressable not found then add it.
```

Comment:
I think should be better to modify the addressable client to add the update function (and maybe also remove function), otherwise shouldn't be possible to update any addressable. What do you think @weichou1229? Can you create this PR in edgex-go?

Reply:
@jduranf in core-metadata, updating Device will also update the associated Addressable, we don't need to make 2 separating calls.

Reply:
@cloudxxx8 updating a Device in core-metadata doesn't update the associated Addressable contents, only updates the Addressable to use.
So, if Addressable name have been changed, then a new Addressable should be added, then Device updated, and finally old Addressable removed. If Addressable name is the same, then just update the Addressable. What do you think?

Reply:
@jduranf You are right.  I misunderstood the code.  However, after discussing with Bruce, we might not want to put the Addressable update logic in Device Cache update.  Users should update addressable through core-metadata Addressable APIs.  The Addressable might be used by other devices or objects.  Updating or removing addressable here might have risk to impact others.  Is it ok with you.
By the way, Bruce will open issue and PR in edgex-go to implement other APIs in Addressable client, such as update.

Reply:
@cloudxxx8 I agree, should be better to update or remove an addressable in other ways. If needed, we can add UpdateAddressable() in devices.go in another PR.
Anyway, if updateAddressable() only creates an addressable if not exists, why no change the name to addAddressableIfNotExist() or something like this?

Reply:
@cloudxxx8 ok, it's fine for me